File: src/my_project/interviews/top_150_questions_round_22/ex_116_kth_largest_element_in_array.py
import random
from typing import List


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        target = len(nums) - k

        def quickselect(left: int, right: int) -> int:
            pivot_index = random.randint(left, right)
            pivot = nums[pivot_index]

            # 3-way partition (Dutch National Flag)
            low = left
            mid = left
            high = right

            while mid <= high:
                if nums[mid] < pivot:
                    nums[low], nums[mid] = nums[mid], nums[low]
                    low += 1
                    mid += 1
                elif nums[mid] > pivot:
                    nums[mid], nums[high] = nums[high], nums[mid]
                    high -= 1
                else:
                    mid += 1

            # All elements equal to pivot are in [low, high]
            if target < low:
                return quickselect(left, low - 1)
            elif target > high:
                return quickselect(high + 1, right)
            else:
                return nums[target]

        return quickselect(0, len(nums) - 1)

File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_116_kth_largest_element_in_array.ts
function findKthLargest(nums: number[], k: number): number {
    const target = nums.length - k;

    function quickselect(left: number, right: number): number {
        const pivotIndex = left + Math.floor(Math.random() * (right - left + 1));
        const pivot = nums[pivotIndex];

        // 3-way partition (Dutch National Flag)
        let low = left;
        let mid = left;
        let high = right;

        while (mid <= high) {
            if (nums[mid] < pivot) {
                [nums[low], nums[mid]] = [nums[mid], nums[low]];
                low++;
                mid++;
            } else if (nums[mid] > pivot) {
                [nums[mid], nums[high]] = [nums[high], nums[mid]];
                high--;
            } else {
                mid++;
            }
        }

        // All elements equal to pivot are in [low, high]
        if (target < low) {
            return quickselect(left, low - 1);
        } else if (target > high) {
            return quickselect(high + 1, right);
        } else {
            return nums[target];
        }
    }

    return quickselect(0, nums.length - 1);
};